packages/hydrogen-react/src/Image.tsx

```diff
@@ -16,205 +21,500 @@ export type ShopifyLoaderOptions = {
+      ...passthroughProps,
     });
+  }
+}
 
-  /* eslint-disable hydrogen/prefer-image-component */
+function unitsMatch(
+  width: string | number = '100%',
+  height: string | number = 'auto',
+): boolean {
   return (
-    <img
-      id={data.id ?? ''}
-      alt={data.altText ?? rest.alt ?? ''}
-      loading={loading ?? 'lazy'}
-      {...rest}
-      src={finalSrc}
-      width={imgElementWidth ?? undefined}
-      height={imgElementHeight ?? undefined}
-      srcSet={finalSrcset}
-      decoding={decoding}
-    />
+    getUnitValueParts(width.toString()).unit ===
+    getUnitValueParts(height.toString()).unit
   );
-  /* eslint-enable hydrogen/prefer-image-component */
+  /*
+      Given:
+        width = '100px'
+        height = 'auto'
+      Returns:
+        false
+
+      Given:
+        width = '100px'
+        height = '50px'
+      Returns:
+        true
+   */
 }
 
-type InternalShopifySrcSetGeneratorsParams = Simplify<
-  ShopifyLoaderOptions & {
-    src: ImageType['url'];
-    widths?: (HtmlImageProps['width'] | ImageType['width'])[];
-    loader?: (params: ShopifyLoaderParams) => string;
+function getUnitValueParts(value: string): {unit: string; number: number} {
+  const unit = value.replace(/[0-9.]/g, '');
+  const number = parseFloat(value.replace(unit, ''));
+
+  return {
+    unit: unit === '' ? (number === undefined ? 'auto' : 'px') : unit,
+    number,
+  };
+  /*
+      Given:
+        value = '100px'
+      Returns:
+        {
+          unit: 'px',
+          number: 100
+        }
+   */
+}
+
+function getNormalizedFixedUnit(value?: string | number): number | undefined {
+  if (value === undefined) {
+    return;
   }
->;
-function internalImageSrcSet({
-  src,
-  width,
-  crop,
-  scale,
-  widths,
-  loader,
-  height,
-}: InternalShopifySrcSetGeneratorsParams): string {
-  const hasCustomWidths = widths && Array.isArray(widths);
-  if (hasCustomWidths && widths.some((size) => isNaN(size as number))) {
-    throw new Error(
-      `<Image/>: the 'widths' must be an array of numbers. Image: ${src}`,
-    );
+
+  const {unit, number} = getUnitValueParts(value.toString());
+
+  switch (unit) {
+    case 'em':
+      return number * 16;
+    case 'rem':
+      return number * 16;
+    case 'px':
+      return number;
+    case '':
+      return number;
+    default:
+      return;
   }
+  /*
+      Given:
+        value = 16px | 1rem | 1em | 16
+      Returns:
+        16
+
+      Given:
+        value = 100%
+      Returns:
+        undefined
+   */
+}
 
-  let aspectRatio = 1;
-  if (width && height) {
-    aspectRatio = Number(height) / Number(width);
+function isFixedWidth(width: string | number): boolean {
+  const fixedEndings = /\d(px|em|rem)$/;
+  return (
+    typeof width === 'number' ||
+    (typeof width === 'string' && fixedEndings.test(width))
+  );
+  /*
+    Given:
+      width = 100 | '100px' | '100em' | '100rem'
+    Returns:
+      true
+  */
+}
+
+export function generateShopifySrcSet(
```

Comment:
@benjaminsehl I'm wondering if there is a reason we wouldn't make this function generic so that it uses the loader they pass in (as a prop) instead of always generating shopify urls. This would still be the default, but if they pass in a loader, we should use it here right? I could be missing something, but if not, I can make this change.

Reply:
Yes you're right — should just be `generateSrcSet`